openverse_catalog/dags/providers/provider_api_scripts/stocksnap.py

```diff
@@ -28,175 +28,165 @@
+    batch_limit = 1000
+    delay = 1  # in seconds
+    headers = {
+        "Accept": "application/json",
+    }
 
+    def __init__(self):
+        super(StockSnapDataIngester, self).__init__()
+        self._page_counter = 1
 
-def main():
-    """
-    This script pulls all the data from the StockSnap and writes it into a
-    .TSV file to be eventually read into our DB.
-    """
-
-    logger.info("Begin: StockSnap script")
-    image_count = _get_items()
-    image_store.commit()
-    logger.info(f"Total images pulled: {image_count}")
-    logger.info("Terminated!")
-
-
-def _get_items():
-    item_count = 0
-    page_number = 1
-    should_continue = True
-    while should_continue:
-        page_endpoint = f"{ENDPOINT}/{page_number}"
-        batch_data = _get_batch_json(endpoint=page_endpoint)
-        if isinstance(batch_data, list) and len(batch_data) > 0:
-            item_count = _process_item_batch(batch_data)
-            page_number += 1
-        else:
-            should_continue = False
-    return item_count
-
-
-def _get_batch_json(
-    endpoint=ENDPOINT, headers=None, retries=RETRIES, query_params=None
-):
-    if headers is None:
-        headers = HEADERS.copy()
-    response_json = delayed_requester.get_response_json(
-        endpoint, retries, query_params, headers=headers
-    )
-    if response_json is None:
+    def get_next_query_params(self):
+        # StockSnap uses /{page} at the end of the endpoint url instead of query params.
+        # Page number is incremented in get_should_continue below.
         return None
-    else:
-        data = response_json.get("results")
-        return data
-
-
-def _process_item_batch(items_batch):
-    for item in items_batch:
-        item_meta_data = _extract_item_data(item)
-        if item_meta_data is None:
-            continue
-        image_store.add_item(**item_meta_data)
-    return image_store.total_items
-
-
-def _extract_item_data(media_data):
-    """
-    Extract data for individual image.
-    """
-    try:
-        foreign_id = media_data["img_id"]
-    except (TypeError, KeyError, AttributeError):
-        return None
-    foreign_landing_url = f"https://{HOST}/photo/{foreign_id}"
-    image_url, width, height = _get_image_info(media_data)
-    if image_url is None:
-        logger.info("Found no image url.")
-        logger.info(f"{json.dumps(media_data, indent=2)}")
-        return None
-    title = _get_title(media_data)
-    if title is None:
-        logger.info("Found no image title.")
-        logger.info(f"{json.dumps(media_data, indent=2)}")
+
+    def get_media_type(self):
+        return "image"
+
+    def endpoint(self):
+        return f"{ENDPOINT_BASE}/{self._page_counter}"
+
+    def get_should_continue(self, response_json):
+        if isinstance(response_json, list) and len(response_json) > 0:
+            self._page_counter += 1
+            return True
+        else:
+            return False
```

Comment:
This works, but I'm not sure `get_should_continue` is the best place to increase the counter 🤔 The intention of this method is to allow additional logic to short-circuit ingestion (so for example, you might place a check for some continue token here). The check here just duplicates the logic in `get_batch_data`, and the counter increment is a side effect that's not really logically related.

We could also increment the counter in `get_next_query_params`. Since it's called before making the first request, you'd have to make sure that it's initialized correctly for the first request. What do you think?

Reply:
:face_palm: That makes total sense. I'll do it.

Reply:
That makes total sense. I'll do it.